### ISSUE TITLE
Does Resoto support multiple AWS accounts?

### ISSUE THREAD

**Issue by iainelder** [someengineering/resoto#648]
### Motivation

This isn't an idea, just a question.

I don't see in the documentation how I would gather data from multiple AWS accounts at once. Is it possible?

### Desired Behavior

-

**Comment by lloesche** [comment]
Yes, there's multiple ways of going about it.

The easiest for demonstration, but not recommended in production(!) is, you give it root account access and then use the `--aws-scrape-org` option on `resotoworker`. This will go into the root account, fetch the list of all sub accounts and collect them all. If not otherwise specified Resoto will use the same config as the [aws cli](https://aws.amazon.com/cli/) to authenticate.

The option `--aws-account-pool-size` determines how many accounts are being collected in parallel. Default is 5. The option `--aws-fork` will make resotoworker fork a collector process per account, which allows you to fully saturate even the largest 64 core servers. The option `--aws-role` allows you to define an IAM role that Resoto is supposed to assume in those accounts.

Every CLI arg can also be supplied via the environment by making it uppercase and prefixing it with the component name. For instance `resotoworker --aws-scrape-org --aws-account-pool-size 50 --aws-fork` can also be turned on by setting:
```
RESOTOWORKER_AWS_SCRAPE_ORG=true
RESOTOWORKER_AWS_FORK=true
RESOTOWORKER_AWS_ACCOUNT_POOL_SIZE=50
```

When running Resoto via Docker a production example using an instance profile and a role that is allowed to assume into every account:
```
RESOTOWORKER_AWS_ACCOUNT_POOL_SIZE=50
RESOTOWORKER_AWS_FORK=true
RESOTOWORKER_AWS_ROLE=Resoto
RESOTOWORKER_AWS_SCRAPE_ORG=true
RESOTOWORKER_COLLECTOR=aws
```

And then inside the resoto container mount a file to `/home/resoto/.aws/config` like
```
[default]
region = us-west-2

role_arn = arn:aws:iam::643255601857:role/Resoto
external_id = g5eMsbsyGDowemdSqpZW3XAMPLExpArw
credential_source = Ec2InstanceMetadata
```



Instead of using auto discovery via the root account you can also specify individual accounts using the `--aws-account 666263750427  775534959047 586712300845` option to have it collect those three account IDs.


The full list of AWS collector plugin options are:
```
  --aws-access-key-id AWS_ACCESS_KEY_ID
                        AWS Access Key ID
  --aws-secret-access-key AWS_SECRET_ACCESS_KEY
                        AWS Secret Access Key
  --aws-role AWS_ROLE   AWS IAM Role
  --aws-role-override   Override any stored roles (e.g. from remote graphs) (default: False)
  --aws-account AWS_ACCOUNT [AWS_ACCOUNT ...]
                        AWS Account
  --aws-region AWS_REGION [AWS_REGION ...]
                        AWS Region (default: all)
  --aws-scrape-org      Scrape the entire AWS Org (default: False)
  --aws-fork            AWS use forked process instead of threads (default: False)
  --aws-scrape-exclude-account AWS_SCRAPE_EXCLUDE_ACCOUNT [AWS_SCRAPE_EXCLUDE_ACCOUNT ...]
                        AWS exclude this Account when scraping the org
  --aws-assume-current  Assume role in current account (default: False)
  --aws-dont-scrape-current
                        Don't scrape current account (default: False)
  --aws-account-pool-size AWS_ACCOUNT_POOL_SIZE
                        AWS Account Thread Pool Size (default: 5)
  --aws-region-pool-size AWS_REGION_POOL_SIZE
                        AWS Region Thread Pool Size (default: 20)
  --aws-collect AWS_COLLECT [AWS_COLLECT ...]
                        AWS services to collect (default: all)
  --aws-no-collect AWS_NO_COLLECT [AWS_NO_COLLECT ...]
                        AWS services not to collect
```


Again, if possible I would recommend against using `--aws-access-key-id`and rather inherit an instance profile. When no auth options are given you can use all options described in https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#environment-variables

The arg based authentication can be useful if you run Resoto outside of AWS, like in an on-prem data center, but would like to collect your AWS infrastructure.

**Comment by lloesche** [comment]
Yes, absolutely, you can use an session policy to restrict access. You could even use temporary credentials but would have to make sure that something renews them out of band. Behind the scenes Resoto uses the [AWS boto3 SDK](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html) and supports all of its authentication options.

If you want all of the Resoto features, including cleanup capabilities it would obviously require read and delete permissions, but if you are only interested in collecting/reading resources, without the ability for Resoto to validate/change tags or clean up resources then a role that only provides read access to the services you are interested in will do just fine.

Side note: by default cleanup is turned off in Resoto and has to be explicitly turned on by providing the `--cleanup` argument. So even if you provide credentials with write/delete permissions Resoto never touches anything without you explicitly turning that code on.

By default `resotoworker` tries to collect all the resources it knows about in all the regions there are. If you know that you're only using e.g. `us-east-1` and `us-west-2` you could provide `--aws-region us-east-1 us-west-2` and it would skip the other regions. Likewise if for instance you're not interested in collecting Cloudformation stacks you could specify `--no-collect cloudformation_stacks` or the other way around, if you're only interested in EC2 instances and volumes you could specify `--collect instances volumes` and Resoto would only collect those and nothing else.

By the way if you run into any issues during setup, feel free to join [our Discord](https://discord.gg/someengineering) and ask me (Lukas) any questions there directly!
